server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -233,12 +285,24 @@ public TemporalAccessor parse(String input) {
      */
     private TemporalAccessor doParse(String input) {
         if (parsers.size() > 1) {
+            Object object = null;
+            DateTimeFormatter lastParsedformatter = null;
             for (DateTimeFormatter formatter : parsers) {
                 ParsePosition pos = new ParsePosition(0);
-                Object object = formatter.toFormat().parseObject(input, pos);
+                object = formatter.toFormat().parseObject(input, pos);
                 if (parsingSucceeded(object, input, pos)) {
-                    return (TemporalAccessor) object;
+                    lastParsedformatter = formatter;
+                    break;
+                }
+            }
+            if (lastParsedformatter != null) {
+                if (canCacheLastParsedFormatter && lastParsedformatter != parsers.get(0)) {
+                    synchronized (parsers) {
```

Comment:
@CaptainDredge it does not look right - the whole iteration part should be synchronized in this case (I am actually not sure what this synchronization block solves to be fair).

Reply:
@reta If we synchronize the whole iteration part then only one thread can read and iterate at a time. What this synchronization blocks solves is the case where the formatter list consistecy is violated i.e. a case where we've remove a formatter on one thread but then other thread read and iterate on it it'll not find the said formatter and will fail to parse. Therefore, we make adding and removing as atomic while other threads reading can iterate on a snapshot of the list provided by CopyOnWriteArrayList

Reply:
> Essentially we needed a way such that multiple threads can read and iterate on the list while only one can write it, where a write operation is `remove + add` atomically

I understand that, the change does not address it, here is the case, 

```
              synchronized (parsers) {
                        parsers.remove(lastParsedformatter); // 1
                        parsers.add(0, lastParsedformatter);  // 2
                    }
```

Since the reader part is NOT synchronized, the other thread could see the parsers list in between step 1 and 2: parser was removed but not added. The  `synchronized` only protect two threads from modification.

Reply:
I see, Yes you're right!! I didn't thought through this case, then the only solution is to have the complete block under a synchronized block. Then this can be solved by classical ReadWriteLock without even using CopyOnWriteArrayList?

Reply:
Exactly, sadly this is the only way to make remove / add atomic

Reply:
@reta see this implementation using read write locks, if this doesn't suffice then I'll resort to synchronizing the whole block which'll decrease concurrency
https://github.com/opensearch-project/OpenSearch/pull/10385